x/denommetadata/ibc_middleware.go

```diff
@@ -0,0 +1,129 @@
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+
+	"github.com/dymensionxyz/dymension-rdk/x/denommetadata/keeper"
+	"github.com/dymensionxyz/dymension-rdk/x/denommetadata/types"
+)
+
+var _ porttypes.IBCModule = &IBCMiddleware{}
+
+// IBCMiddleware implements the ICS26 callbacks for the transfer middleware
+type IBCMiddleware struct {
+	porttypes.IBCModule
+	keeper keeper.Keeper
+}
+
+// NewIBCMiddleware creates a new IBCMiddleware given the keeper and underlying application
+func NewIBCMiddleware(k keeper.Keeper, app porttypes.IBCModule) IBCMiddleware {
+	return IBCMiddleware{
+		IBCModule: app,
+		keeper:    k,
+	}
+}
+
+// OnRecvPacket registers the denom metadata if it does not exist
+func (im IBCMiddleware) OnRecvPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	relayer sdk.AccAddress,
+) exported.Acknowledgement {
+	packetData := new(transfertypes.FungibleTokenPacketData)
+	if err := types.ModuleCdc.UnmarshalJSON(packet.GetData(), packetData); err != nil {
+		err = errorsmod.Wrapf(errortypes.ErrInvalidType, "cannot unmarshal ICS-20 transfer packet data")
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	if len(packetData.Memo) == 0 {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	rawDenom := packetData.Denom
+
+	denomTrace := transfertypes.ParseDenomTrace(rawDenom)
+	if denomTrace.Path == "" {
+		denomTrace.Path = fmt.Sprintf("%s/%s", packet.GetDestPort(), packet.GetDestChannel())
+	}
+
+	if im.keeper.HasDenomMetadata(ctx, denomTrace) {
+		return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+	}
+
+	if err := im.createNewDenom(ctx, packetData, denomTrace); err != nil {
+		return channeltypes.NewErrorAcknowledgement(err)
+	}
+
+	return im.IBCModule.OnRecvPacket(ctx, packet, relayer)
+}
+
+func (im IBCMiddleware) createNewDenom(ctx sdk.Context, packetData *transfertypes.FungibleTokenPacketData, denomTrace transfertypes.DenomTrace) error {
+	packetMetaData, err := types.ParsePacketMetadata(packetData.Memo)
+	if errors.Is(err, types.ErrMemoUnmarshal) || errors.Is(err, types.ErrMemoDMEmpty) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	dm := packetMetaData.DenomMetadata
+
+	denomUnits := make([]*banktypes.DenomUnit, 0, len(dm.DenomUnits))
+	for _, du := range dm.DenomUnits {
+		if du.Exponent == 0 {
+			continue
+		}
+		ndu := &banktypes.DenomUnit{
+			Denom:    du.Denom,
+			Exponent: du.Exponent,
+			Aliases:  du.Aliases,
+		}
+		denomUnits = append(denomUnits, ndu)
+	}
+
+	newDenomMetadata := types.DenomMetadata{
+		TokenMetadata: banktypes.Metadata{
+			Description: dm.Description,
+			DenomUnits:  denomUnits,
+			Base:        denomTrace.IBCDenom(),
+			Display:     dm.Display,
+			Name:        dm.Name,
+			Symbol:      dm.Symbol,
+			URI:         dm.URI,
+			URIHash:     dm.URIHash,
+		},
+		DenomTrace: denomTrace.GetFullDenomPath(),
+	}
+
+	if err := im.keeper.CreateDenomMetadata(ctx, newDenomMetadata); err != nil {
```

Comment:
@zale144 how this data is trusted in this stage?
anyone can send any token from the hub and add any memo, no?

Reply:
@mtsitrin Yes, good point. As far as I understand, the only place an IBC transfer can come to the rollapp is from the hub. So we should make sure that the user cannot send a transfer from the hub to the rollapp with the memo having a field `"denom_metadata"`, which is the one carrying the token metadata payload.
 
So my idea is that on the hub side we don't allow the `"denom_metadata"` field in the memo to be set by anyone except the `SendPacket` middleware. 
Like here https://github.com/dymensionxyz/dymension/blob/478335bbbd4f01e3c04cf948520053707f3e5620/x/transferinject/types/packet_metadata.go#L54

Another idea I had was to sign the `"denom_metadata"` payload on the hub and validate the signature on the rollapp. This would mean introducing a private key to the hub and having all the rollapps be deployed with a copy of its public key counterpart. But I believe that in the IBC protocol there is already a mechanism for proving the validity of the IBC packet, so I thought it would probably be redundant to add the signature scheme.

Again, my understanding might not be complete, so it would be great if this part gets reviewed more strictly. An audit would perhaps also be useful, in case we miss something.

Reply:
let's discuss it over an ADR.
We need to validate we didn't miss any edge cases/attack vectors

Reply:
Did you guys have further discussion? Can you link the adr in the issue?

Reply:
>This would mean introducing a private key to the hub and having all the rollapps be deployed with a copy of its public key counterpart.

The hub cannot sign anything. The hub is public code only
Only validators can sign something
Any block is already signed by the validators